### PR TITLE
Add tokenizer option to allow Matcher handling for all rules

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Add tokenizer option `with_faster_rules_heuristics` that determines whether the special cases applied by the internal `Matcher` are filtered by whether they contain affixes or space. If `True` (default), the rules are filtered to prioritize speed over rare edge cases. If `False`, all rules are included in the final `Matcher`-based pass over the doc.

### Types of change

Enhancement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
Something is going very wrong in the new test in the first python 3.6 test, but I am running it again to see if it's consistently very wrong.

**Comment by svlandeg** [comment]
Oof, that is one scary looking failing test :|

**Comment by adrianeboyd** [comment]
I can reproduce it locally, too. The last commit is not going to fix it.

File: website/docs/api/tokenizer.md [hunk 1]
```diff
@@ -44,15 +44,16 @@ how to construct a custom tokenizer with different tokenization rules, see the
...
+| `suffix_search`                | A function matching the signature of `re.compile(string).search` to match suffixes. ~~Optional[Callable[[str], Optional[Match]]]~~                                            |
+| `infix_finditer`               | A function matching the signature of `re.compile(string).finditer` to find infixes. ~~Optional[Callable[[str], Iterator[Match]]]~~                                            |
+| `token_match`                  | A function matching the signature of `re.compile(string).match` to find token matches. ~~Optional[Callable[[str], Optional[Match]]]~~                                         |
+| `url_match`                    | A function matching the signature of `re.compile(string).match` to find token matches after considering prefixes and suffixes. ~~Optional[Callable[[str], Optional[Match]]]~~ |
+| `with_faster_rules_heuristics` | Whether to restrict the final `Matcher`-based pass for rules to those containing affixes or space. Defaults to `True`. ~~bool~~                                               |
```
Comment (svlandeg) [inline website/docs/api/tokenizer.md:56]: Add that this is new in 3.3 ?

**Comment by adrianeboyd** [comment]
I'm still not entirely sure about the naming here...

**Comment by svlandeg** [comment]
> I'm still not entirely sure about the naming here...

If we want to have something shorter, I think either `with` or maybe even `rules` could be removed from the name?

**Comment by adrianeboyd** [comment]
Yeah, I think `faster_heuristics` would be okay. Matt also suggested `prioritize_speed` or something along those lines.

**Comment by svlandeg** [comment]
Let's settle on `faster_heuristics`?

**Comment by honnibal** [comment]
Agree that the name is the least worst option 🙂 👍

**Pull request merged**